### ISSUE TITLE
SimpleInjector lock exception when used with SqlPersistence

### ISSUE THREAD

**Issue by Gachapen** [WilliamBZA/NServicebus.SimpleInjector#12]
When using SqlPersistence, Simpleinjector throws an InvalidOperationException when starting the endpoint. This happens because this library tries to register `SqlPersistenceInstaller` in `SimpleInjectorObjectBuilder.Configure` after the SimpleInjector container was locked by `SimpleInjectorObjectBuilder.Build` (triggered by feature setups).

You don't actually need to configure SqlPersistence to reproduce the problem, it happens by only including the nuget. It should be resolved by registering `SqlPersistenceInstaller` before the feature setups, but I can't do this because it is internal. I'm not sure if it is a problem with some logic in this project, or if SqlPersistence needs to be changed.

The exception thrown is as follows:
```
The container can't be changed after the first call to GetInstance, GetAllInstances, Verify, and some calls of GetRegistration. Please see https://simpleinjector.org/locked to understand why the container is locked. The following stack trace describes the location where the container was locked:

   at NServiceBus.ObjectBuilder.SimpleInjector.SimpleInjectorObjectBuilder.Build(Type typeToBuild) in NServicebus.SimpleInjector\src\NServiceBus.SimpleInjector\SimpleInjectorObjectBuilder.cs:line 78
   at NServiceBus.CommonObjectBuilder.Build[T]() in NServiceBus\src\NServiceBus.Core\ObjectBuilder\Common\CommonObjectBuilder.cs:line 28
   at NServiceBus.Features.ReceiveFeature.<>c.<Setup>b__1_0(IBuilder b) in NServiceBus\src\NServiceBus.Core\Pipeline\Incoming\ReceiveFeature.cs:line 20
   at NServiceBus.Pipeline.PipelineSettings.<>c__DisplayClass8_0`1.<Register>b__0(IBuilder b) in NServiceBus\src\NServiceBus.Core\Pipeline\PipelineSettings.cs:line 132
   at NServiceBus.Pipeline.RegisterStep.CreateBehavior(IBuilder defaultBuilder) in NServiceBus\src\NServiceBus.Core\Pipeline\RegisterStep.cs:line 143
   at NServiceBus.Pipeline`1.<>c__DisplayClass0_0.<.ctor>b__0(RegisterStep r) in NServiceBus\src\NServiceBus.Core\Pipeline\Pipeline.cs:line 25
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at NServiceBus.Pipeline`1..ctor(IBuilder builder, PipelineModifications pipelineModifications) in NServiceBus\src\NServiceBus.Core\Pipeline\Pipeline.cs:line 24
   at NServiceBus.InitializableEndpoint.CreateReceiveComponent(ReceiveConfiguration receiveConfiguration, TransportInfrastructure transportInfrastructure, QueueBindings queueBindings, IPipelineCache pipelineCache, EventAggregator eventAggregator) in NServiceBus\src\NServiceBus.Core\InitializableEndpoint.cs:line 153
   at NServiceBus.InitializableEndpoint.<Initialize>d__1.MoveNext() in NServiceBus\src\NServiceBus.Core\InitializableEndpoint.cs:line 65
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at NServiceBus.InitializableEndpoint.Initialize()
   at NServiceBus.Endpoint.Create(EndpointConfiguration configuration) in NServiceBus\src\NServiceBus.Core\Endpoint.cs:line 18
   at NServiceBus.Endpoint.<Start>d__1.MoveNext() in NServiceBus\src\NServiceBus.Core\Endpoint.cs:line 27
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at NServiceBus.Endpoint.Start(EndpointConfiguration configuration)
```

With stack trace:
```
   at SimpleInjector.Container.ThrowWhenContainerIsLockedOrDisposed()
   at SimpleInjector.Container.AddRegistrationInternal(Type serviceType, Registration registration)
   at SimpleInjector.Container.AddRegistration(Type serviceType, Registration registration)
   at NServiceBus.ObjectBuilder.SimpleInjector.SimpleInjectorObjectBuilder.Configure(Type component, DependencyLifecycle dependencyLifecycle) in NServicebus.SimpleInjector\src\NServiceBus.SimpleInjector\SimpleInjectorObjectBuilder.cs:line 115
```
```

**Comment by Gachapen** [comment]
Actually, manually registering `SqlPersistenceInstaller` does not resolve the issue, because it then throws on `GetExistingCollectionRegistrationsFor` because `collectionRegistrations` was set to `null` during `Build`.

**Comment by Gachapen** [comment]
I found a way to solve this issue, but it involves modifying NServiceBus.Core.

The core issue here is that `NServiceBus.Features.ReceiveFeature` is setup before the installers are run. Because `ReceiveFeature`'s setup function builds some dependencies, the SimpleInjector becomes locked, and the installers can't be registered in the nest step.

My solution is to move the configuration/registration of the installers before the receive feature setup:
```
            if (shouldRunInstallers)
            {
                var username = GetInstallationUserName();
                ConfigureInstallers(concreteTypes, username);
            }

            var receiveComponent = CreateReceiveComponent(receiveConfiguration, transportInfrastructure, queueBindings, pipelineCache, eventAggregator);

            if (shouldRunInstallers)
            {
                var username = GetInstallationUserName();

                if (settings.CreateQueues())
                {
                    await receiveComponent.CreateQueuesIfNecessary(queueBindings, username).ConfigureAwait(false);
                }

                await RunInstallers(concreteTypes, username).ConfigureAwait(false);
            }
```
```
        void ConfigureInstallers(IEnumerable<Type> concreteTypes, string username)
        {
            foreach (var installerType in concreteTypes.Where(t => IsINeedToInstallSomething(t)))
            {
                container.ConfigureComponent(installerType, DependencyLifecycle.InstancePerCall);
            }
        }

        async Task RunInstallers(IEnumerable<Type> concreteTypes, string username)
        {
            foreach (var installer in builder.BuildAll<INeedToInstallSomething>())
            {
                await installer.Install(username).ConfigureAwait(false);
            }
        }
```
While not ideal, this solves the issue. I can't see a way to solve this in the SimpleInjector library, so maybe we should submit an issue to NServiceBus?

**Comment by WilliamBZA** [comment]
We could also [re-introduce the ability to clone a container](https://github.com/WilliamBZA/NServicebus.SimpleInjector/pull/6/commits/5644e820945410ddc32308e5079fb55182312d6e#diff-d8cd224ce5fe3b5b8efe43be8ac7b5b4L21) into a new one. I really don't like that though.

I'll dig around and see if I can come up with something clever...

**Comment by WilliamBZA** [comment]
I raised an issue in core and in the meantime allowed cloning of the container again. It's horrible and has a performance penalty, but until core changes it is the only way to get it running.

Released in [3.0.3](https://www.nuget.org/packages/NServiceBus.SimpleInjector/3.0.3).

**Comment by timbussmann** [comment]
hey @Gachapen 
Just wanted to let you know that we have fixed this with NServiceBus 7.1.5 and 7.0.3. Here is our release announcement if you want more details: https://discuss.particular.net/t/nservicebus-7-1-5-and-7-0-3-patch-releases-available/881

**Comment by Gachapen** [comment]
Great! Thank you, both of you. It works well now.
